trips/consumers.py

```diff
@@ -0,0 +1,149 @@
+from channels.generic.websocket import AsyncJsonWebsocketConsumer
+from channels.db import database_sync_to_async
+
+from trips.serializers import NestedTripSerializer, TripSerializer
+from trips.models import Trip
+
+
+class TaxiConsumer(AsyncJsonWebsocketConsumer):
+
+    @database_sync_to_async
+    def _get_user_group(self, user):
+        return user.groups.first().name
+
+    @database_sync_to_async
+    def _create_trip(self, data):
+        serializer = TripSerializer(data=data)
+        serializer.is_valid(raise_exception=True)
+        return serializer.create(serializer.validated_data)
+
+    @database_sync_to_async
+    def _get_trip_data(self, trip):
+        return NestedTripSerializer(trip).data
+
+    @database_sync_to_async
+    def _get_trip_ids(self, user):
+        user_groups = user.groups.values_list('name', flat=True)
+
+        if 'driver' in user_groups:
+            trip_ids = user.trips_as_driver.exclude(
+                status=Trip.COMPLETED
+            ).only('id').values_list('id', flat=True)
+        else:
+            trip_ids = user.trips_as_rider.exclude(
+                status=Trip.COMPLETED
+            ).only('id').values_list('id', flat=True)
+
+        return map(str, trip_ids)
+
+    @database_sync_to_async
+    def _update_trip(self, data):
+        instance = Trip.objects.get(id=data.get('id'))
+        serializer = TripSerializer(data=data)
+        serializer.is_valid(raise_exception=True)
+        return serializer.update(instance, serializer.validated_data)
+
+    async def connect(self):
+        user = self.scope['user']
+
+        if user.is_anonymous:
+            await self.close()
+        else:
+            user_group = await self._get_user_group(user)
+
+            if user_group == 'driver':
+                await self.channel_layer.group_add(
+                    group='drivers',
+                    channel=self.channel_name
+                )
+
+            for trip_id in await self._get_trip_ids(user):
+                await self.channel_layer.group_add(
+                    group=trip_id,
+                    channel=self.channel_name
+                )
+
+            await self.accept()
+
+    async def disconnect(self, code):
+        user = self.scope['user']
+
+        if user.is_anonymous:
+            await self.close()
+
+        else:
+            user_group = await self._get_user_group(user)
+            if user_group == 'driver':
+                await self.channel_layer.group_discard(
+                    group='drivers', channel=self.channel_name
+                )
+
+                for trip_id in await self._get_trip_ids(user):
```

Comment:
Fix this indentation.... This code should run, regardless of whether a user is a driver or rider .. Currently it only runs for drivers